image/qcow2/qcow2.go

```diff
@@ -962,6 +962,114 @@ func readZero(p []byte, off int64, sz uint64) (int, error) {
 	return l, err
 }
 
+// clusterStatus returns an extent describing a single cluster. off must be aligned to
+// cluster size.
+func (img *Qcow2) clusterStatus(off int64) (image.Extent, error) {
+	var cm clusterMeta
+	if err := img.getClusterMeta(off, &cm); err != nil {
+		return image.Extent{}, err
+	}
+
+	if !cm.Allocated {
+		// If there is no backing file, or the cluster cannot be in the backing file,
+		// return an unallocated cluster.
+		if img.backingImage == nil || off >= img.backingImage.Size() {
+			// Unallocated cluster reads as zeros.
+			unallocated := image.Extent{Start: off, Length: int64(img.clusterSize), Zero: true}
+			return unallocated, nil
+		}
+
+		// Get the cluster from the backing file.
+		length := int64(img.clusterSize)
+		if off+length > img.backingImage.Size() {
+			length = img.backingImage.Size() - off
+		}
+		parent, err := img.backingImage.Next(off, length)
+		if err != nil {
+			return parent, err
+		}
+		// The backing image may be a raw image not aligned to cluster size.
+		parent.Length = int64(img.clusterSize)
+		return parent, nil
+	}
+
+	// Cluster present in this image.
+	allocated := image.Extent{
+		Start:      off,
+		Length:     int64(img.clusterSize),
+		Allocated:  true,
+		Compressed: cm.Compressed,
+		Zero:       cm.Zero,
```

Comment:
Add comment about extended l2 - the cluster is always non-zero. We need to iterate on the sub cluster allocation bitmap to find which subcluster are allocated/zero. Current code is correct but less efficient - we will do very course grain (1 MiB) zero detection for extended l2, instead of handling 32 subclusters in every cluster.

Reply:
Commented in getClusterMeta.